azure-cli: fix installing extensions externally

impl opened the pull request:
## Description of changes

For those of us not using the new azure-cli-extensions mechanism to manage extensions, it is not possible to install or update extensions unless pip is on the `PYTHONPATH`.

This used to be the case by default, but a relatively recent change to either Python packaging (likely 6c85fff302615c62bf4f632bca661bc48298b0a3, as identified in other issues) or construction of the PYTHONPATH in this package (aacf05daec3141ce2bb34fd7c021a86401ac8c51) removed it.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

katexochen commented:
> This used to be the cause by default, but a relatively recent change to Python packaging (likely [6c85fff](https://github.com/NixOS/nixpkgs/commit/6c85fff302615c62bf4f632bca661bc48298b0a3), as identified in other issues) removed it.

I think the reason why pip is missing could also be https://github.com/NixOS/nixpkgs/pull/276695/commits/aacf05daec3141ce2bb34fd7c021a86401ac8c51. Previously we would just wrap the binary with the path of our build environment, which likely contains pip.

katexochen commented on pkgs/tools/admin/azure-cli/default.nix:
```diff
@@ -207,6 +207,7 @@ py.pkgs.toPythonApplication (py.pkgs.buildAzureCliPackage rec {
     packaging
     paramiko
     pbr
+    pip
```
Can we add pip conditionally in case `withImmutableConfig` is false?

impl committed:
azure-cli: fix installing extensions externally

For those of us not using the new azure-cli-extensions mechanism to
manage extensions, it is not possible to install or update extensions
unless pip is on the PYTHONPATH.

This used to be the case by default, but a relatively recent change to
either Python packaging (likely
6c85fff302615c62bf4f632bca661bc48298b0a3, as identified in other issues)
or construction of the PYTHONPATH in this package
aacf05daec3141ce2bb34fd7c021a86401ac8c51 removed it.

impl commented:
Thanks for the quick feedback @katexochen! I made those changes and validated that it works as intended:

```
palmier % nix build .#azure-cli
warning: Git tree '/home/impl/p/projects/nix/nixpkgs' is dirty

palmier % ./result/bin/az extension add --name ssh
Default enabled including preview versions for extension installation now. Disabled in May 2024. Use '--allow-preview true' to enable it specifically if needed. Use '--allow-preview false' to install stable version only. 
An error occurred. Pip failed with status code 1. Use --debug for more information.

palmier % nix build "$(nix eval .#azure-cli --raw --apply 'azure-cli: (azure-cli.override { withImmutableConfig = false; }).drvPath')^*"
warning: Git tree '/home/impl/p/projects/nix/nixpkgs' is dirty

palmier % ./result/bin/az extension add --name ssh
Default enabled including preview versions for extension installation now. Disabled in May 2024. Use '--allow-preview true' to enable it specifically if needed. Use '--allow-preview false' to install stable version only. 

palmier % ./result/bin/az ssh --help

Group
    az ssh : SSH into resources (Azure VMs, Arc servers, etc) using AAD issued openssh certificates.
        SSH into Arc-enabled servers without requiring a public IP address or SSH into Azure Virtual
        Machines. AAD issued openssh certificates for authentication currently only supported for
        Linux.

Commands:
    arc    : SSH into Azure Arc Servers.
    cert   : Create an SSH RSA certificate signed by AAD.
    config : Create an SSH config for resources (Azure VMs, Arc Servers, etc) which can then be used
             by clients that support OpenSSH configs and certificates.
    vm     : SSH into Azure VMs or Arc Servers.

To search AI knowledge base for examples, use: az find "az ssh"
```

katexochen reviewed (approved):
Thanks!

I tested that installing the mutable extension works with

```
nixpkgs on  azure-cli-add-pip [$] 
❯ nix build --impure --expr 'with (import ./. {}); azure-cli.override { withImmutableConfig = false; }'

nixpkgs on  azure-cli-add-pip [$] 
❯ ./result/bin/az extension list
[]

nixpkgs on  azure-cli-add-pip [$] 
❯ ./result/bin/az extension add --name ssh
Default enabled including preview versions for extension installation now. Disabled in May 2024. Use '--allow-preview true' to enable it specifically if needed. Use '--allow-preview false' to install stable version only. 
 - Installing ..
nixpkgs on  azure-cli-add-pip [$] took 5s 
❯ ./result/bin/az extension list
[
  {
    "experimental": false,
    "extensionType": "whl",
    "name": "ssh",
    "path": "/home/katexochen/.azure/cliextensions/ssh",
    "preview": false,
    "version": "2.0.3"
  }
]
```

And afterwards confirmed the immutable version still does the expected with

```
nixpkgs on  azure-cli-add-pip [$] 
❯ nix build --impure --expr 'with (import ./. {}); azure-cli.withExtensions [ azure-cli.extensions.aks-preview ]'

nixpkgs on  azure-cli-add-pip [$] 
❯ ./result/bin/az extension list
[
  {
    "experimental": false,
    "extensionType": "whl",
    "name": "aks-preview",
    "path": "/nix/store/ymggggqr6hckb6rycb4xjc4gvx99v071-azure-cli-extensions/aks-preview",
    "preview": true,
    "version": "1.0.0b5"
  }
]
```
